README.md
CKEditor 5 video feature
========================================

[![npm version](https://badge.fury.io/js/%40visao%2Fckeditor5-video.svg)](https://www.npmjs.com/package/@visao/ckeditor5-video)

This package implements the video feature for CKEditor 5. The feature is introduced in a granular form implemented by a couple of plugins.
It was strongly inspired from the ckeditor5-image package.

## Demo

- See `sample/app.js`

## Documentation

## Installation
Add this to your custom build or inside your project.

- With npm

`npm install --save-dev @visao/ckeditor5-video`


-With yarn

`yarn add -D @visao/@visao/ckeditor5-video    `
- Works pretty much just like Image upload. 

## Plugins

#### Video Plugin 
- Plugin to parse videos in the editor
- Mandatory for the other plugins VideoRelated plugins
    
```
ClassicEditor
    .create( document.querySelector( '#editor' ), {
        plugins: [Video],
        video: {}
    } )

```

#### VideoUpload Plugin 
- Plugin to upload video files via toolbar upload prompt or drag and drop functionalities 
- Specify allowed media(mime) types. Default => `['mp4', 'webm', 'ogg']`
- Allow multiple file upload or not, Default => `true`
- Add the `videoUpload` toolbar option to access the file repository 
- Must provide an `UploadAdapter`. 
See [ckeditor5 documentation](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/upload-adapter.html)
- The use of the Video plugin is mandatory for this plugin to work
    
```
function VideoUploadAdapterPlugin( editor ) {
    editor.plugins.get('FileRepository').createUploadAdapter = (loader) => {
        return new VideoUploadAdapter(loader);
    };
}

ClassicEditor
    .create( document.querySelector( '#editor' ), {
        plugins: [Video, VideoUpload],
        extraPlugins: [VideoUploadAdapterPlugin],
        toolbar: ['videoUpload'],
        video: {
            upload: {
                types: ['mp4'],
                allowMultipleFiles: false,
            }
        }
    } )
```

#### VideoToolbar Plugin
- Balloon toolbar for different Video plugin plugins
- See VideoResizing and VideoStyle sections for examples

#### VideoResizing Plugin 
- Plugin for resizing the video in the editor
- Should work just like image resize. See the ck-editor 5 documentation for more examples.
```
ClassicEditor
    .create( document.querySelector( '#editor' ), {
        plugins: [Video, VideoToolbar, VideoResize] or [Video, VideoToolbar, VideoResizeEditing, VideoResizeHandles],
        video: {
            resizeUnit: 'px'
            // Configure the available video resize options.
            resizeOptions: [
                {
                    name: 'videoResize:original',
                    value: null,
                    label: 'Original',
                    icon: 'original'
                },
                {
                    name: 'videoResize:50',
                    value: 50,
                    label: '50',
                    icon: 'medium'
                },
                {
                    name: 'videoResize:75',
                    value: '75',
                    label: '75',
                    icon: 'large'
                }
            ],
            toolbar: [
                'videoResize',
                '|',
                'videoResize:50',
                'videoResize:75',
                'videoResize:original'
            ]
        },
    } )
```

#### VideoStyle Plugin
- Plugin for styling the video plugins.
- Should work just like image resize. See the ck-editor 5 documentation for more examples.
- Predefined styles are:
  - `full`
  - `side`
  - `alignLeft`
  - `alignCenter`
  - `alignRight`
```
ClassicEditor
    .create( document.querySelector( '#editor' ), {
        plugins: [Video, VideoToolbar, VideoStyle]
        video: {
            styles: [
                'alignLeft', 'alignCenter', 'alignRight'
            ],
            toolbar: ['videoStyle:alignLeft', 'videoStyle:alignCenter', 'videoStyle:alignRight']
        },
    } )
```
  


## License

Licensed under the terms of 
[GNU General Public License Version 2 or later](http://www.gnu.org/licenses/gpl.html). For full details about the license, 
please check the `LICENSE.md` file.


CHANGELOG.md
Changelog
=========

Changes for the past releases below.

## 29.1.0(2021-09-04)
### Upgraded
* CkEditor5 dependencies to v29.1.0

## 29.0.1(2021-09-04)
### Upgraded
* Bugfix for when inserting a video, the wrong file vs files options is being used

## 29.0.0(2021-08-22)
### Upgraded
* CkEditor5 dependencies to v29.0.0

## 28.0.0(2021-06-12)
### Upgraded
* CkEditor5 dependencies to v28.0.0

## 27.1.0(2021-06-01)
### Upgraded
* CkEditor5 dependencies to v27.1.0
* Adding VideoInsert to the lib

## 23.1.0(2020-11-19)
### Upgraded
* CkEditor5 dependencies to v23.1.0

## 23.0.1(2020-11-19)
### Added
* Chinese translations

## 23.0.0(2020-11-18)
### Added
* Style support [#1](https://github.com/Technologie-Visao/ckeditor5-video/issues/1) 
* Balloon toolbar support
* More documentation

### Fixed
* Jumping versions to follow the ck-editor5 dependencies
of this package. (23.0.x here)

## 0.0.2(2020-10-21)
### Added
* A bit of doc in the readme

### Fixed
* Uncommented UploadVideoCommand toolbar ui condition 


## 0.0.1(2020-10-20)
### Added
* Published package to npm
* Video plugin
* Video Upload Plugin
* Video Resize Plugin


sample/app.js
import ClassicEditor from '@ckeditor/ckeditor5-editor-classic/src/classiceditor';
import Paragraph from '@ckeditor/ckeditor5-paragraph/src/paragraph';
import Bold from '@ckeditor/ckeditor5-basic-styles/src/bold';
import Italic from '@ckeditor/ckeditor5-basic-styles/src/italic';
import CKEditorInspector from '@ckeditor/ckeditor5-inspector';
import List from '@ckeditor/ckeditor5-list/src/list';
import Heading from '@ckeditor/ckeditor5-heading/src/heading';
import Essentials from "@ckeditor/ckeditor5-essentials/src/essentials";
import VideoUpload from "../src/videoupload";
import Video from "../src/video";
import VideoResize from "../src/videoresize";
import VideoToolbar from "../src/videotoolbar";
import VideoStyle from "../src/videostyle";
import VideoInsert from "../src/videoinsert";

class VideoUploadAdapter {
    constructor( loader ) {
        this.loader = loader;
    }

    upload() {
        const uploadVideo = async (file) => {
            this.loader.uploaded = false;
            return new Promise((resolve) => {
                setTimeout(() => {
                    this.loader.uploaded = true;
                    resolve({ default: 'https://file-examples-com.github.io/uploads/2017/04/file_example_MP4_480_1_5MG.mp4' });
                }, 2000);
            });
        };

        return this.loader.file.then((file) => uploadVideo(file));
    }

    abort() {
        return Promise.reject();
    }
}

function VideoUploadAdapterPlugin( editor ) {
    editor.plugins.get('FileRepository').createUploadAdapter = (loader) => {
        return new VideoUploadAdapter(loader);
    };
}

ClassicEditor
    .create( document.querySelector( '#editor' ), {
        plugins: [ Essentials, Paragraph, Bold, Italic, Heading, List, VideoToolbar, Video, VideoUpload, VideoResize, VideoStyle, VideoInsert ],
        extraPlugins: [VideoUploadAdapterPlugin],
        toolbar: [ 'heading', 'bold', 'italic', 'numberedList', 'bulletedList', 'videoUpload' ],
        video: {
            upload: {
                types: ['mp4'],
                allowMultipleFiles: false,
            },
            styles: [
                'alignLeft', 'alignCenter', 'alignRight'
            ],

            // Configure the available video resize options.
            resizeOptions: [
                {
                    name: 'videoResize:original',
                    label: 'Original',
                    icon: 'original'
                },
                {
                    name: 'videoResize:50',
                    label: '50',
                    icon: 'medium'
                },
                {
                    name: 'videoResize:75',
                    label: '75',
                    icon: 'large'
                }
            ],

            // You need to configure the video toolbar, too, so it shows the new style
            // buttons as well as the resize buttons.
            toolbar: [
                'videoStyle:alignLeft', 'videoStyle:alignCenter', 'videoStyle:alignRight',
                '|',
                'videoResize:50',
                'videoResize:75',
                'videoResize:original'
            ]
        }
    } )
    .then( editor => {
        CKEditorInspector.attach( editor );

        window.editor = editor;
    } )
    .catch( error => {
        console.error( error.stack );
    } );
